### PR TITLE
Data Prepper expressions - Set operator fix

### PR DESCRIPTION
**Pull request by kkondaka**
### Description
Set operator 'in' and 'not in' are not working as expected. This change fixes this.
 
### Issues Resolved
Resolves #[Issue number to be closed when this PR is merged]
 
### Check List
- [ X] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [ X] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dlvenable** [comment]
@kkondaka , This should resolve #3854 correct?

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/OpenSearchSink.java [hunk 1]
```diff
@@ -448,9 +448,9 @@ public void doOutput(final Collection<Record<Event>> records) {
       }
 
       SerializedJson serializedJsonNode = null;
-      if (StringUtils.equals(action, OpenSearchBulkActions.UPDATE.toString()) ||
-          StringUtils.equals(action, OpenSearchBulkActions.UPSERT.toString()) ||
-          StringUtils.equals(action, OpenSearchBulkActions.DELETE.toString())) {
+      if (StringUtils.equals(eventAction, OpenSearchBulkActions.UPDATE.toString()) ||
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/OpenSearchSink.java:451]: This appears unrelated and appears to be from another PR. Please remove this change.

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/GenericExpressionEvaluator_ConditionalIT.java [hunk 2]
```diff
@@ -214,6 +214,19 @@ private static Stream<Arguments> validExpressionArguments() {
                 Arguments.of("/sourceIp != null", event("{\"sourceIp\": [10, 20]}"), true),
                 Arguments.of("/sourceIp == null", event("{\"sourceIp\": [\"test\", \"test_two\"]}"), false),
                 Arguments.of("/sourceIp == null", event("{\"sourceIp\": {\"test\": \"test_two\"}}"), false),
+                Arguments.of("/value in {200.222, 300.333, 400}", event("{\"value\": 400.0}"), true),
```
Comment (dlvenable) [inline data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/GenericExpressionEvaluator_ConditionalIT.java:217]: Please use `arguments`. See https://github.com/opensearch-project/data-prepper/pull/4798/files#diff-720b03bb55dc916760c0e8106c53d4fd6422c00bc31feb0cb9a90f24ea66fb4e

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeCoercionService.java [hunk 3]
```diff
@@ -76,6 +79,55 @@ public Object coercePrimaryTerminalNode(final TerminalNode node, final Event eve
                     return longValue;
                 }
                 return Integer.valueOf(nodeStringValue);
+            case DataPrepperExpressionParser.SetInitializer:
+                String[] setMembers = nodeStringValue.trim().substring(1,nodeStringValue.length()-1).split(",");
```
Comment (dlvenable) [inline data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeCoercionService.java:83]: This case statement is too complicated already. Let's move this out.

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeCoercionService.java [hunk 4]
```diff
@@ -76,6 +79,55 @@ public Object coercePrimaryTerminalNode(final TerminalNode node, final Event eve
...
+                    if (s.length() > 1 && s.charAt(0) == '"' && s.charAt(s.length()-1) == '"') {
+                        stringMemberCount++;
+                    } else if (s.contains(".")) {
+                        floatMember = true;
+                    }
+                }
+                if (stringMemberCount > 0 && stringMemberCount != setMembers.length) {
+                    throw new RuntimeException("All set members should be of same type");
+                }
+                if (stringMemberCount == setMembers.length) {
```
Comment (dlvenable) [inline data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeCoercionService.java:97]: Casting the types is something we do elsewhere for other literals right? We should remove this code in favor of using the code that supports existing literals.

File: data-prepper-expression/src/main/antlr/DataPrepperExpression.g4 [hunk 5]
```diff
@@ -181,6 +186,13 @@ literal
     | Null
     ;
 
+fragment
+SetLiteral
```
Comment (dlvenable) [inline data-prepper-expression/src/main/antlr/DataPrepperExpression.g4:190]: I don't think we want another fragment here. We should use the other `literal` and remove this.

File: data-prepper-expression/src/main/antlr/DataPrepperExpression.g4 [hunk 6]
```diff
@@ -121,8 +121,13 @@ regexPattern
...
 
-setInitializer
-    : LBRACE primary (SET_DELIMITER primary)* RBRACE
+SetInitializer
+    : LBRACE SetMembers RBRACE
+    ;
+
+fragment
+SetMembers
+    : SetLiteral (SPACE* COMMA SPACE* SetLiteral)*
```
Comment (dlvenable) [inline data-prepper-expression/src/main/antlr/DataPrepperExpression.g4:130]:
```suggestion
    : literal (SPACE* COMMA SPACE* literal)*
```

Let's use the same literal. Also, can we use `primary` as it used to be used? Perhaps sets of sets would be appropriate.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeCoercionService.java (hunk 4 above)
Comment (kkondaka) [reply data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeCoercionService.java:97]: The code doesn't get called for each of the set members. Let me know if I am missing something.

File: data-prepper-expression/src/main/antlr/DataPrepperExpression.g4 (hunk 5 above)
Comment (kkondaka) [reply data-prepper-expression/src/main/antlr/DataPrepperExpression.g4:190]: I felt doing something like `/boolean_value in {true, false}` doesn't make sense. Similarly, I felt checking for `null` in a set is not that useful

File: data-prepper-expression/src/main/antlr/DataPrepperExpression.g4 (hunk 6 above)
Comment (kkondaka) [reply data-prepper-expression/src/main/antlr/DataPrepperExpression.g4:130]: Maybe in future? I do not want to complicate this any further for now. We have documented this since the beginning without supporting it.

File: data-prepper-expression/src/main/antlr/DataPrepperExpression.g4 (hunk 5 above)
Comment (dlvenable) [reply data-prepper-expression/src/main/antlr/DataPrepperExpression.g4:190]: Maybe it's not very useful, but the literals may grow as well. I think if we just have a literal, we can have a nice recursive structure. This is also similar to other programming languages.

**Review by dlvenable — APPROVED** [review]
Nice, this looks great!

**Review by graytaylor0 — APPROVED** [review]
